v8.2.0 raises `Can't instantiate abstract class` when a test class implements an abstract class

Before, we could have a test class inheriting from an ABC (and unittest.TestCase), now it errors because pytest is trying to instantiate the ABC itself. Running pytest should only instantiate the child test class (which indeed implements all the abstract methods of its ABC parent).
This issue does not appear in v8.1.2 and was likely introduced in https://github.com/pytest-dev/pytest/pull/12089.

Minimal example:
```python
from abc import ABC, abstractmethod
from unittest import TestCase


class Base(TestCase, ABC):
    @abstractmethod
    def foo(self): ...


class Tests(Base):

    def foo(self): ...

    def test(self): ...
```

Output:

```
(test) tests 🔥pytest test.py    
============================================================================= test session starts ==============================================================================
platform darwin -- Python 3.9.19, pytest-8.2.0, pluggy-1.5.0
rootdir: ...
configfile: pytest.ini
plugins: env-1.1.3
collected 1 item / 1 error                                                                                                                                                     

==================================================================================== ERRORS ====================================================================================
________________________________________________________________ ERROR collecting test.py ________________________________________________________________
/Users/martin/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
/Users/martin/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/runner.py:389: in collect
    return list(collector.collect())
/Users/martin/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/unittest.py:90: in collect
    self.session._fixturemanager.parsefactories(self.newinstance(), self.nodeid)
/Users/martin/miniconda3/envs/test/lib/python3.9/site-packages/_pytest/unittest.py:75: in newinstance
    return self.obj("runTest")
E   TypeError: Can't instantiate abstract class Base with abstract method foo
=========================================================================== short test summary info ============================================================================
ERROR test.py::Base - TypeError: Can't instantiate abstract class Base with abstract method foo
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=============================================================================== 1 error in 0.04s ===============================================================================
```

pip list:

```
Package        Version
-------------- -------
exceptiongroup 1.2.1
iniconfig      2.0.0
packaging      24.0
pip            23.3.1
pluggy         1.5.0
pytest         8.2.0
pytest-env     1.1.3
setuptools     68.2.2
tomli          2.0.1
wheel          0.41.2
```
Thanks for your work in general, it's a great package.

--- a/changelog/12275.bugfix.rst
+++ b/changelog/12275.bugfix.rst
@@ -1,0 +1,1 @@
+Fix collection error upon encountering an :mod:`abstract <abc>` class, including abstract `unittest.TestCase` subclasses.

--- a/src/_pytest/python.py
+++ b/src/_pytest/python.py
@@ -368,7 +368,11 @@ class PyCollector(PyobjMixin, nodes.Collector, abc.ABC):
             return False
 
     def istestclass(self, obj: object, name: str) -> bool:
-        return self.classnamefilter(name) or self.isnosetest(obj)
+        if not (self.classnamefilter(name) or self.isnosetest(obj)):
+            return False
+        if inspect.isabstract(obj):
+            return False
+        return True
 
     def _matches_prefix_or_glob_option(self, option_name: str, name: str) -> bool:
         """Check if the given name matches the prefix or glob-pattern defined

--- a/src/_pytest/unittest.py
+++ b/src/_pytest/unittest.py
@@ -1,6 +1,7 @@
 # mypy: allow-untyped-defs
 """Discover and run std-library "unittest" style tests."""
 
+import inspect
 import sys
 import traceback
 import types
@@ -49,13 +50,18 @@ if TYPE_CHECKING:
 def pytest_pycollect_makeitem(
     collector: Union[Module, Class], name: str, obj: object
 ) -> Optional["UnitTestCase"]:
-    # Has unittest been imported and is obj a subclass of its TestCase?
     try:
+        # Has unittest been imported?
         ut = sys.modules["unittest"]
+        # Is obj a subclass of unittest.TestCase?
         # Type ignored because `ut` is an opaque module.
         if not issubclass(obj, ut.TestCase):  # type: ignore
             return None
     except Exception:
+        return None
+    # Is obj a concrete class?
+    # Abstract classes can't be instantiated so no point collecting them.
+    if inspect.isabstract(obj):
         return None
     # Yes, so let's collect it.
     return UnitTestCase.from_parent(collector, name=name, obj=obj)

--- a/testing/python/collect.py
+++ b/testing/python/collect.py
@@ -262,6 +262,32 @@ class TestClass:
         result = pytester.runpytest()
         assert result.ret == ExitCode.NO_TESTS_COLLECTED
 
+    def test_abstract_class_is_not_collected(self, pytester: Pytester) -> None:
+        """Regression test for #12275 (non-unittest version)."""
+        pytester.makepyfile(
+            """
+            import abc
+
+            class TestBase(abc.ABC):
+                @abc.abstractmethod
+                def abstract1(self): pass
+
+                @abc.abstractmethod
+                def abstract2(self): pass
+
+                def test_it(self): pass
+
+            class TestPartial(TestBase):
+                def abstract1(self): pass
+
+            class TestConcrete(TestPartial):
+                def abstract2(self): pass
+            """
+        )
+        result = pytester.runpytest()
+        assert result.ret == ExitCode.OK
+        result.assert_outcomes(passed=1)
+
 
 class TestFunction:
     def test_getmodulecollector(self, pytester: Pytester) -> None:

--- a/testing/test_unittest.py
+++ b/testing/test_unittest.py
@@ -1640,3 +1640,31 @@ def test_raising_unittest_skiptest_during_collection(
     assert skipped == 1
     assert failed == 0
     assert reprec.ret == ExitCode.NO_TESTS_COLLECTED
+
+
+def test_abstract_testcase_is_not_collected(pytester: Pytester) -> None:
+    """Regression test for #12275."""
+    pytester.makepyfile(
+        """
+        import abc
+        import unittest
+
+        class TestBase(unittest.TestCase, abc.ABC):
+            @abc.abstractmethod
+            def abstract1(self): pass
+
+            @abc.abstractmethod
+            def abstract2(self): pass
+
+            def test_it(self): pass
+
+        class TestPartial(TestBase):
+            def abstract1(self): pass
+
+        class TestConcrete(TestPartial):
+            def abstract2(self): pass
+        """
+    )
+    result = pytester.runpytest()
+    assert result.ret == ExitCode.OK
+    result.assert_outcomes(passed=1)

